File: Deprerated/Siren/Attribute/SirenClassAttribute.cs
// Copyright (c) 2015 fjz13. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.
using System;

namespace Siren.Attribute
{
    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct | AttributeTargets.Enum)]
    public class SirenClassAttribute : System.Attribute
    {
        public Type Template { get; set; }
        public string Directory { get; set; }
        public SirenGenerateMode Mode { get; set; }
        public string NewHeader { get; set; }
        public string EnumUnderType { get; set; }


        public bool IsEmbeded
        {
            get { return Mode.HasFlag(SirenGenerateMode.Embeded); }
        }

        public SirenClassAttribute(Type template, string directory, SirenGenerateMode mode = SirenGenerateMode.Generate)
        {
            Template = template;
            Directory = directory;
            Mode = mode;

            if (!Mode.HasFlag(SirenGenerateMode.Embeded)&& !Mode.HasFlag(SirenGenerateMode.Generate) &&!Mode.HasFlag(SirenGenerateMode.Suppress))
            {
                Mode |= SirenGenerateMode.Generate;
            }
        }
    }

}

File: Medusa/MedusaCore/Core/Siren/Schema/Attribute/SirenClassAttribute.cpp
// Copyright (c) 2015 fjz13. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.
#include "MedusaCorePreCompiled.h"
#include "SirenClassAttribute.h"
#include "Core/IO/Stream/IStream.h"
#include "Core/Log/Log.h"

MEDUSA_BEGIN;

SirenClassAttribute::SirenClassAttribute(SirenClassGenerateMode mode, const StringRef& dir)
	:mMode(mode), mDir(dir)
{

}

SirenClassAttribute::~SirenClassAttribute(void)
{
	
}





bool SirenClassAttribute::OnLoaded()
{
	if (mKeyValues.RemoveKey("Generate"))
	{
		MEDUSA_FLAG_ADD(mMode, SirenClassGenerateMode::Generate);
	}

	if (mKeyValues.RemoveKey("Suppress"))
	{
		MEDUSA_FLAG_ADD(mMode, SirenClassGenerateMode::Suppress);
		MEDUSA_FLAG_REMOVE(mMode, SirenClassGenerateMode::Generate);

	}

	if (mKeyValues.RemoveKey("Embeded"))
	{
		MEDUSA_FLAG_ADD(mMode, SirenClassGenerateMode::Embeded);
	}

	if (mKeyValues.RemoveKey("SuppressCompare"))
	{
		MEDUSA_FLAG_ADD(mMode, SirenClassGenerateMode::SuppressCompare);
	}

	if (mKeyValues.RemoveKey("SuppressCopy"))
	{
		MEDUSA_FLAG_ADD(mMode, SirenClassGenerateMode::SuppressCopy);
	}

	if (mKeyValues.RemoveKey("SirenConfig"))
	{
		MEDUSA_FLAG_ADD(mMode, SirenClassGenerateMode::SirenConfig);
	}

	if (mKeyValues.Has("Dir"))
	{
		mDir = mKeyValues.Get("Dir");
		mKeyValues.RemoveKey("Dir");
	}

	return true;
}


bool SirenClassAttribute::LoadFrom(IStream& stream)
{
	RETURN_FALSE_IF_FALSE(ISirenAttribute::LoadFrom(stream));
	mMode = stream.Read<SirenClassGenerateMode>();
	mDir = stream.ReadString();
	return true;
}

bool SirenClassAttribute::SaveTo(IStream& stream) const
{
	RETURN_FALSE_IF_FALSE(ISirenAttribute::SaveTo(stream));
	stream.Write(mMode);
	stream.WriteString(mDir);
	return true;
}

MEDUSA_END;

